### ISSUE TITLE
Proposal: Approximately-equal operator ~=

### ISSUE THREAD

**Issue by Jehoel** [dotnet/csharplang#1675]
### Motivation:

The existing Equals `==` operator is overridden for types for common comparison scenarios. However the nature of "equality" in .NET (and software in general) is often more nuanced. For example I often find myself needing to perform ordinal case-insensitive string comparisons, but the `==(String,String)` operator invokes `String.Equals(String,String)` which performs an ordinal, case-sensitive, string comparison - this means I can't use the nice `==` shorthand syntax but need to replace them with the verbose `String.Equals( x, y, StringComparison.OrdinalIgnoreCase )`.

I'd like to propose that the C# language adds a new operator `~=` (or `~~` or `=~`) which has the semantics of an "approximately equal" comparison.

1. The primary use-case being to invoke an ordinal, case-insensitive comparison when used with `String` operands - however it has applications for other types too:
1. For IEEE-754 types, where `==` often doesn't work as-intended, the `~=` operator could be used to check for a near-enough comparison without needing to repeat the equality-within-tolerance code in the `Double.Equals` [documentation page](https://msdn.microsoft.com/en-us/library/ya2zha7s(v=vs.110).aspx).

An obvious shortcoming is the necessity to parameterize the operation: while ordinal case-insensitive string equality is straightforward, it's an anglocentric presumption of mine that it has greater utility than an ordinal, accent-insensitive string equality (for many other languages) or even more general: [a Soundex string comparison](https://en.wikipedia.org/wiki/Soundex). This parameterization could be done by treating the `~=` operator as a ternary operator (with 3 operands), the parameter being specified with some kind of separator after the second operand (e.g. `:`).

For example:

    bool floatApproxEquals = ( 0.3333f ~= 1f/3f ); // binary operator with default tolerance of 0.01
    bool floatApproxEquals = ( 0.3333f ~= 1f/3f : 0.001f ); // ternary, explicit tolerance of 0.001
    bool stringApproxEquals = ( "abc" ~= "ABC" ); // binary operator with default argument CompareOptions.IgnoreCase
    bool stringApproxEquals = ( "abc" ~= "ABC" : CompareOptions.IgnoreNonSpace ); // ternary, explicit CompareOptions value

The operator overload signature would look like:

    public class Foo
    {
        // Binary
        public static Boolean operator~=( Foo first, Foo second )
        {
            // ...
        }

        // Ternary 
        public static Boolean operator~=( Foo first, Foo second, FooComparisonOptions mode )
        {
            // ...
        }
    }

As the type of the third operand is statically defined the compiler can enforce this, so doing ` 0.3f ~= 1f : CompareOptions.None` would trigger a compile-time error because the `~=` operator for `Single` and `Double` would expects a floating-point `tolerance` value.

### Augmentation for Linq for SQL:

As a bonus, this works well with Linq when used against a SQL data-source where `==(String,String)` is actually a case-insensitive comparison (according to most DBMS collations), so by adopting `~=` the Linq implementation could specify the precisely required `COLLATE` argument to textual comparison operations (and perhaps restore `==` to stricter ordinal, case-sensitive, semantics shared with C#):

    dbContext.People.Where( person => person.Name == "Anders Hejlsberg" ).ToList();
    // SELECT ... WHERE Name = "Anders Hejlsberg" COLLATE Latin1_General_CS_AS

    dbContext.People.Where( person => person.Name ~= "Anders Hejlsberg"  ).ToList();
    // SELECT ... WHERE Name = "Anders Hejlsberg" COLLATE Latin1_General_CI_AI

    dbContext.People.Where( person => person.Name ~= "Anders Hejlsberg" : CompareOptions.IgnoreNonSpace  ).ToList();
    // SELECT ... WHERE Name = "Anders Hejlsberg" COLLATE Latin1_General_CS_AI

**Comment by CyrusNajmabadi** [comment]
The moment you need to parameterize an operator... well.. then an operator probably isn't the right tool for the job.

**Comment by Mafii** [comment]
Well written proposal. Really. But I disagree that we need an operator for this.

**Comment by Jehoel** [comment]
I'm sober now, closing :)

**Comment by mikenakis** [comment]
The way I am thinking of implementing this, without any support from the language, is as follows:

(warning: I have not tried it yet!) 

1. Introduce a `struct Approx` which contains a double value and a double tolerance.
2. Define the normal '`==`' operator between `double` and `Approx` to do an approximate comparison.

So, assuming `Approx` has these functions:

```
public static Approx Of( double value, double tolerance ) => new Approx( value, tolerance );
public static Approx Of( double value ) => new Approx( value, Constants.DefaultDoubleTolerance );
```

the code you end up writing looks like this:

```
if( someValue == Approx.Of( 5 ) ) ...
if( a == Approx.Of( b, 1e-4 ) ) ...
```

In the above example I dealt only with `double`, but the mechanism that I have described can be used to implement special ways of comparing for any data type, and with the help of C# 11 static abstract methods it should be possible to even write it only once for all numeric types.

However, support from the language would be nice. So, I created this: [[Proposal] Add more overloadable operators #10226](https://github.com/dotnet/csharplang/discussions/10226)
